rfcs/0000-error-handling.md

```diff
@@ -0,0 +1,153 @@
+
+Less catastrophic errors do have a downside: It's harder to notice that anything is wrong at all. We'll include an error in the graphql response for unhealthy and failed subgraphs. We will improve and document the indexing status API, so that developers can programmatically check details on a subgraph's health and build their own alerts and dashboards.
+
+## Urgency
+
+Sooner is better than later.
+
+## Terminology
+
+A new subgraph health status called "unhealthy" is added, so instead of a subgraph being failed or not, it can be healthy, unhealthy or failed.
+
+## Detailed Design
+
+### Error handling
+
+Subgraphs now have three possible health status: 'Healthy' is normal operation, 'Unhealthy' is syncing with errors, and 'Failed' is halted due to an error.
+
+If a WASM handler terminates with a trap, the subgraph becomes unhealthy. The node discards all entity operations and data source operations done by the handler and continues syncing.
+
+For subgraphs that are not yet synced, any handler error is still a failure and halts syncing, so errors in pending versions will not make their way to the current version. 
+
+When the node starts, we currently reset the failed flag on subgraphs. Failed subgraphs will still be reset to healthy to attempt to recover, but unhealthy subgraphs are kept unhealthy, since the subgraph will already be past the errors.
+
+Any graphql queries to an unhealthy or failed subgraph will include an error of the form:
+
+```
+{
+	message: "Subgraph not healthy"
+}
+```
+
+So that applications will always be aware that queries may have inconsistent data. Developers should refer to the indexing status API for details on the errors.
+
+### Indexing Status API
+
+The indexing status API will be revamped.
+
+The following top level query is added:
+
+    indexingStatusForSubgraphName(subgraphName: String!): SubgraphIndexingStatus
+
+Which returns null if the name is not found, and returns the status for the current version of the subgraph otherwise.
+
+A field `lastHealthyBlock`  is added to record the block previous to the first error.
+
+Errors are made more than just a string with the `Error` type.
+
+The complete indexing status API schema is:
+
+```graphql
+type Query {
+  indexingStatusForSubgraphName(subgraphName: String!): SubgraphIndexingStatus
+  indexingStatusesForSubgraphName(subgraphName: String!): [SubgraphIndexingStatus!]!
+  indexingStatuses(subgraphs: [String!]): [SubgraphIndexingStatus!]!
+}
+
+type SubgraphIndexingStatus {
+  subgraph: String!
+  synced: Boolean!
+  health: Health!
+  errors: [Error!]! # Sorted from first to last
```

Comment:
Quick comment about the errors: there may be a _lot_ of them. A few ideas:

1. Only keep the most recent N errors around, for a fixed N (e.g. 100). I think this is bad, because you'll usually want to know where things started to go wrong and how. The first 100 would be similarly limiting.

2. Implement pagination a la `first`/`skip`. The main problem with this is that every new block may add new errors, so while you're paging through the errors, new ones will be added and the pages may jump around confusingly. What could work is starting `first` at the first/oldest error. Then you'd get stable pages and you'll get the most useful errors first, without needing to paginate.

Thoughts?

Reply:
I hadn't considered this problem. Paginating from the oldest error sounds good. This might not need a change in the RFC but it's something to keep in mind for the implementation.

Reply:
A third option would be to move the errors into separate `indexingErrorsForSubgraph` resolvers to which you have to pass the blocks for each network involved in the subgraph(currently only the Ethereum block hash) and mix that with pagination from the most recent error. Tying it to a specific block would make pagination stable, but starting from the most recent error is less useful. Approach 2 has my vote, I think.

Reply:
@leoyvens Yeah, agreed.

Reply:
What's an upper bound on the number of errors we might encounter? Number of handlers * number of blocks? I assume dynamic data sources play into this, too, in which case each block could generate a ton of errors (thousands) We should predefine an error budget and if a subgraph goes beyond that still fail it. I don't think it helps anybody if we have subgraphs that can generate and store huge numbers at every block - they'll most likely be useless and just eat up resources.

Reply:
We can limit the amount of errors stored, though I'd imagine that a successful handler will likely generate more data than an erroed one. I'm not in favor of having an error budget after which the subgraph is halted, thousands of operations could fail and there could still be useful ones going through.

Reply:
I am concerned with what that will do to the hosted service: right now, subgraphs that are clearly useless and fail do not cause any work after starting and failing them again. With this change, any subgraph, no matter how useless it is will continue to cause work. Currently, we have 686 assigned deployments that are working, and 112 that are failed. We would therefore increase the load in hosted by 16% for no real purpose whatsoever - if the failed subgraphs were important, they would have gotten fixed.

Reply:
Yes, keeping subgraphs with errors running will require resources. To avoid that jump, we can make currently failed subgrahps stay failed when doing the migration.

We can establish some criteira by which we stop indexing an unhealthy subgraph on the hosted service, and possibly make that criteira built-in to the graph node.

For example, if every single handler errors for a configured number of blocks, then we stop indexing the subgraph. But because we don't yet have data on how many subgraphs would actually be stopped on a given criteria, I feel it's premature to establish one. If we run this on the hosted service for a while, we'll be better informed and able to experiment with criteira for failing unhealthy subgraphs.

Reply:
1. I think it's fair to limit the number of errors we store to, say, 1000. And drop pagination even. As long as we also log the caught errors so we can establish a full list from the logs if needed.

2. The extra resources to keep failed subgraphs running are a good reason to make the feature opt-in, and keep the current failing behavior as the default. There are a few things we can do to prevent _everyone_ from using it, and thereby keep the resource use at bay, like stressing that this should only be used for production subgraphs. Warning about it in graph-cli. Things like that.

Reply:
It's true that not every subgraph needs this, and disabling it by default might save some hosted service resources.

The tradeoff being that some subgraph developers will find out the hard way that they need this, by going through an outage before finding out in our support channels that their issues can be mitigated by enabling this.

Reply:
As long as we document that

* it is only recommended for production subgraphs
* but it _is_ recommended for production subgraphs

clear enough, I think that's ok. It could be good to have a "best practices for production subgraphs" section in the docs.

Reply:
Alright, then we need to decide how this is specified in the manifest. For simplicity this can be a top-level flag on the manifest, applying to all data sources, though we could make it per-data source if we want to. Some brainstorming:

- Have a general `production` boolean, which we could possibly use for other behaviour that differs between a debug mode and a production mode.

- Have a more specific flag that reflects this particular behaviour. I'm not sure how to describe it, maybe `failOnMappingError: false`.